### PR TITLE
Allow reading of default parameters from @ROMFS/defaults.parm

### PR DESCRIPTION
**Pull request by peterbarker**
This does work - defaults can come from the ROMFS file.  I've tested that on hardware.

Embedding defaults like this means you won't run out of space for parameters until you run out of flash, and saves the wasted space the "parameter space" typically represents (1k on small boards, 8kB on large boards).  Also applies to AP_Periph builds, which often cut down the available space but still waste that space.

I'm just not sure what to do about backwards compatibility.  If we merge this then you can no longer take a `.apj` and "tack on" default parameters using `apj_tool`.

### CODE REVIEW

**Comment by peterbarker** [comment]
... I guess one point is that this uses up flash space.  If the parameter you want to set a value for has an (e.g. `AP_FOO_BAR_DEFAULT` then you're better off using that flash-wise, but this PR will allow you to fetch the defaults file from ROMFS via ftp, which is nice

File: libraries/AP_HAL_ChibiOS/Util.h [hunk 1]
```diff
@@ -39,6 +39,11 @@ class ChibiOS::Util : public AP_HAL::Util {
     bool run_debug_shell(AP_HAL::BetterStream *stream) override { return false; }
     uint32_t available_memory() override;
 
+    // get path to custom defaults file for AP_Param
+    const char* get_custom_defaults_file() const override {
+        return "@ROMFS/defaults.parm";
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Util.h:44]: I'd like to see this logged to bin logs (added to the list)

**Review by tridge — CHANGES_REQUESTED** [review]
i'd like to understand the CPU cost and memory cost of this - looking at the resulting data structure

File: libraries/AP_HAL_ChibiOS/Util.h (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_HAL_ChibiOS/Util.h:44]: Done

**Comment by peterbarker** [comment]
> i'd like to understand the CPU cost and memory cost of this - looking at the resulting data structure

The resulting data structure is identical between "embedded param defaults" and "load defaults file" (apj-tacked-on-parameters and defaults.parm in ROMFS).  Both methods allocate a `param_defaults` array based on the number of detected parameters (search for `new param_override`)

**Comment by peterbarker** [comment]
Include files were tested on CubeOrangePlus / CubeOrangePlus-bdshot:

This is the -bdshot variant, which `@include`s the non-bdshot variant defaults.parm file:
```
MANUAL> Getting @ROMFS/defaults.parm as -
# setup correct defaults for battery monitoring for cube power brick
BATT2_CURR_PIN 4
BATT2_VOLT_PIN 13
BATT_AMP_PERVLT 39.877
BATT_VOLT_MULT 12.02
BATT2_AMP_PERVLT 39.877
BATT2_VOLT_MULT 12.02
# setup ADSB
ADSB_TYPE 1
SERIAL5_BAUD 57
SERIAL5_PROTOCOL 1
```

**Comment by peterbarker** [comment]
- improved error message from apj_tool
 - added option to build server to enable the parameter area

**Review by tridge — APPROVED** [review]

**Pull request merged**